### DISCUSSION TITLE
Interpretation of `REPRESENT` string format.

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#176] [Q&A]
I'm looking for clarification concerning the string of significant digits returned in the given buffer to `REPRESENT`.  Typically I'd expect from C printf() family a string formatting `123.45` would typically yield for `%e` a string `1.23450e02`.

But as I read `REPRESENT` the phrase in bold implies that string returned would be without a decimal point, ie. `123450` with `S: 3 FALSE TRUE`.  Is this a correct interpretation?

From 12.6.1.2143 `REPRESENT`:

> The character string shall consist of the `u` most significant digits of the significand represented as a decimal fraction with the ***implied decimal point to the left of the first digit***, and the first digit zero only if all digits are zero. 

Test cases to illustrate what I understand:

```forth
6 SET-PRECISION

t{ 0.1e PAD PRECISION REPRESENT S" 100000" PAD PRECISION COMPARE -> 0 FALSE TRUE 0 }t
t{ 0.4e PAD PRECISION REPRESENT S" 400000" PAD PRECISION COMPARE -> 0 FALSE TRUE 0 }t
t{ 0.6e PAD PRECISION REPRESENT S" 600000" PAD PRECISION COMPARE -> 0 FALSE TRUE 0 }t

t{    0.0   PAD PRECISION REPRESENT S" 000000" PAD PRECISION COMPARE -> 0 FALSE TRUE 0 }t
t{  123.45  PAD PRECISION REPRESENT S" 123450" PAD PRECISION COMPARE -> 3 FALSE TRUE 0 }t
t{ -123.45  PAD PRECISION REPRESENT S" 123450" PAD PRECISION COMPARE -> 3  TRUE TRUE 0 }t
t{  0.12345 PAD PRECISION REPRESENT S" 123450" PAD PRECISION COMPARE -> 0 FALSE TRUE 0 }t
t{ -0.12345 PAD PRECISION REPRESENT S" 123450" PAD PRECISION COMPARE -> 0  TRUE TRUE 0 }t

2.718281828459 FCONSTANT e
3.141592653589 FCONSTANT pi
6.62607015e-34 FCONSTANT planck

t{ e        PAD PRECISION REPRESENT S" 271828" PAD PRECISION COMPARE -> 1  FALSE TRUE 0 }t
t{ pi       PAD PRECISION REPRESENT S" 314159" PAD PRECISION COMPARE -> 1  FALSE TRUE 0 }t
t{ planck   PAD PRECISION REPRESENT S" 662607" PAD PRECISION COMPARE -> -33 FALSE TRUE 0 }t
```

**Comment by ruv** [comment]
>  string returned would be without a decimal point

Yes, the string returned by `represent` does not contain a decimal point, as well as a minus sign (if applicable).

An interesting question is what exponent value it should return in the following cases:
  - the first input parameter on the data stack is zero;
  - the input floating point value is zero.

I think, when the first input parameter is zero, the floating-point number should be rounded to the nearest <del>integer</del> <ins>rationale</ins> that is a [power of ten](https://en.wikipedia.org/wiki/Power_of_10), and the returned parameter _n_ is that power exponent.

**Reply by SirWumpus** [reply]
I think there is a problem with the wording about `caddr` `u`.  Every where else the pair `caddr` `u`, the `u` is the length of a string or the size of a buffer.  Expressing `u` in terms in the size of the significand is a little confusing.  If `u` is zero then the buffer `caddr` should be empty; not sure `n` has any meaning without at least the first digit of the significand in that case.

In the case of `0.0` should the buffer contain just `0` or padded with zeros eg. `00...0u`; how many zero's do you need to make a point (outside of politics).

I think `REPRESENT` seriously needs test cases to demonstrate what is expected.

**Reply by ruv** [reply]
>  If _u_ is zero then the buffer _caddr_ should be empty;

_c-addr_ is an address, it cannot be empty.  The buffer itself is empty in the sense that its length is zero.

> not sure _n_ has any meaning without at least the first digit of the significand in that case.

In this case _n_ could mean the decimal order of the number.  Namely, if rounding to nearest is used,  then _n_ (which is a member of the data type **n**) is the value that minimizes the function `f(n) = abs( 10^n - abs(r) )`.

-----

> In the case of `0.0` should the buffer contain just 0 or padded with zeros eg. 00...0u;

According to the spec, the buffer shall be padded with zeros: "The character string shall consist of the _u_ most significant digits".

-----

> I think `REPRESENT` seriously needs test cases to demonstrate what is expected.

Yes, sure!
